src/SkeletalTrapezoidationGraph.cpp

```diff
@@ -0,0 +1,323 @@
+    {
+        node_locator.emplace(&*node_it, node_it);
+    }
+    
+    auto safelyRemoveEdge = [this, &edge_locator](edge_t* to_be_removed, std::list<edge_t>::iterator& current_edge_it, bool& edge_it_is_updated)
+    {
+        if (current_edge_it != edges.end()
+            && to_be_removed == &*current_edge_it)
+        {
+            current_edge_it = edges.erase(current_edge_it);
+            edge_it_is_updated = true;
+        }
+        else
+        {
+            edges.erase(edge_locator[to_be_removed]);
+        }
+    };
+
+    auto should_collapse = [snap_dist](node_t* a, node_t* b) 
+    { 
+        return shorterThen(a->p - b->p, snap_dist); 
+    };
+        
+    for (auto edge_it = edges.begin(); edge_it != edges.end();)
+    {
+        if (edge_it->prev)
+        {
+            edge_it++;
+            continue;
+        }
+        
+        edge_t* quad_start = &*edge_it;
+        edge_t* quad_end = quad_start; while (quad_end->next) quad_end = quad_end->next;
+        edge_t* quad_mid = (quad_start->next == quad_end)? nullptr : quad_start->next;
+
+        bool edge_it_is_updated = false;
+        if (quad_mid && should_collapse(quad_mid->from, quad_mid->to))
+        {
+            assert(quad_mid->twin);
+            int count = 0;
+            for (edge_t* edge_from_3 = quad_end; edge_from_3 && edge_from_3 != quad_mid->twin; edge_from_3 = edge_from_3->twin->next)
+            {
+                edge_from_3->from = quad_mid->from;
+                edge_from_3->twin->to = quad_mid->from;
+                if (count > 50)
+                {
+                    std::cerr << edge_from_3->from->p << " - " << edge_from_3->to->p << '\n';
+                }
+                if (++count > 1000) 
+                {
+                    break;
+                }
+            }
+
+            // o-o > collapse top
+            // | |
+            // | |
+            // | |
+            // o o
+            if (quad_mid->from->some_edge == quad_mid)
+            {
+                if (quad_mid->twin->next)
+                {
+                    quad_mid->from->some_edge = quad_mid->twin->next;
+                }
+                else
+                {
+                    quad_mid->from->some_edge = quad_mid->prev->twin;
+                }
+            }
+            
+            nodes.erase(node_locator[quad_mid->to]);
+
+            quad_mid->prev->next = quad_mid->next;
+            quad_mid->next->prev = quad_mid->prev;
+            quad_mid->twin->next->prev = quad_mid->twin->prev;
+            quad_mid->twin->prev->next = quad_mid->twin->next;
+
+            safelyRemoveEdge(quad_mid->twin, edge_it, edge_it_is_updated);
+            safelyRemoveEdge(quad_mid, edge_it, edge_it_is_updated);
+        }
+
+        //  o-o
+        //  | | > collapse sides
+        //  o o
+        if ( should_collapse(quad_start->from, quad_end->to) && should_collapse(quad_start->to, quad_end->from))
```

Comment:
Can we nest this if-statement inside the previous one? See my previous comment here: https://github.com/Ultimaker/CuraEngine/pull/1210#discussion_r392099762

Reply:
I don't think so: `should_collapse` could be false for the top edge and be true for the start/end distances independently. 

I _think_ it may even be commented on below that last if-block:
```
// If only one side had collapsable length then the cell on the other side of that edge has to collapse
// if we would collapse that one edge then that would change the quad_start and/or quad_end of neighboring cells
// this is to do with the constraint that !prev == !twin.next
```